indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java

```diff
@@ -281,31 +286,25 @@ public void unannounceSegments(Iterable<DataSegment> segments) throws IOExceptio
       }
     };
 
-    // NOTE: getVersion will block if there is lock contention, which will block plumber.getSink
-    // NOTE: (and thus the firehose)
+    // NOTE: getVersion will block (and thus block the firehose) if there is lock contention
 
     // Shouldn't usually happen, since we don't expect people to submit tasks that intersect with the
     // realtime window, but if they do it can be problematic. If we decide to care, we can use more threads in
     // the plumber such that waiting for the coordinator doesn't block data processing.
-    final VersioningPolicy versioningPolicy = new VersioningPolicy()
-    {
-      @Override
-      public String getVersion(final Interval interval)
-      {
-        try {
-          // Side effect: Calling getVersion causes a lock to be acquired
-          final LockAcquireAction action = new LockAcquireAction(TaskLockType.EXCLUSIVE, interval, lockTimeoutMs);
-          final TaskLock lock = Preconditions.checkNotNull(
-              toolbox.getTaskActionClient().submit(action),
-              "Cannot acquire a lock for interval[%s]",
-              interval
-          );
+    final VersioningPolicy versioningPolicy = interval -> {
```

Comment:
`AppenderatorDriver` uses an `ActionBasedAllocator` (https://github.com/druid-io/druid/pull/5261/files#diff-917318f978c8bd3d86decbfdc0166c59R660) which internally acquires a lock for the interval of the allocated segment. If this versioningPolicy is for acquiring a lock, it is not needed.

BTW, I could see that your motivation was to support user-defined versioningPolicy in RealtimeIndexTask (https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/druid-user/AATjuCUFS0E/36OIMaJjAAAJ). This also makes me feel that we need to remove this.

Reply:
I left the versioningPolicy as it was with the previous incarnation of `RealtimeIndexTask`, the only change I made was to clean up the anonymous inner class by converting it to a lambda.

I'm happy to remove this if it isn't needed but I didn't feel comfortable removing it without the background on why it was there in the first place

Reply:
Any Druid's index tasks should acquire a lock for a pair of (dataSource, interval) before writing some data to a segment. This is to prevent multiple tasks from writing for the same interval at the same time. 

`RealtimeIndexTask` was using a `VersioningPolicy` which is responsible for acquiring a lock for the given interval. This might make sense because the segment version is decided from the acquired lock. However, I feel it's sort of hacky because this `VersioningPolicy` encompasses some side effects like communicating with overlords. 

In addition, this `VersioningPolicy` internally uses `LockAcquireAction` which always acquires a lock of the new version. This means, if `VersioningPolicy` is called for a late arrival, `LockAcquireAction` returns a new version for the interval of that late arrival, thereby eventually overshadowing all other segments ingested to the same interval so far. `SegmentAllocateAction` can handle this problem properly and returns a new segmentId for late arrivals which doesn't overshadow other segments ingested by the same task.

BTW, does this PR also targets to support ingestion of late arrivals? I guess it does from this discussion (https://github.com/druid-io/druid/pull/4774#issuecomment-344621045). If so, please add a test for it.